Chargeback project date only

kbrock opened the pull request:
### Before

When chargebacks are run for date only, the data is grouped by date, and across all projects. So displaying a project is not correct.

### After

Project is no longer displayed.
The Vm was correctly removed from the report for the vm date-only reports.
Made the changes for the other chargeback reports to properly remove those columns.

Fryguy commented:
@kbrock Would is be possible to write specs, or do we already have them?

Fryguy commented:
I'm having trouble following the logic.  The static_cols, from reading the old code, started with report_static_cols, then specific fields were removed in certain cases.  However, in the new code, the static_cols is never set to report_static_cols.

Separately I noticed the rpt.cols order changed - not sure that matters or not as long as cols_order is the same, but I am concerned it could introduce issues elsewhere?

kbrock commented:
update:
- changed order of the `rpt.cols`

@Fryguy look at the first commit. It should be identical to the original code, just with a case statement.

Then the second commit shows what I modified. And I think my assumptions are much more clear. I'm hoping you can vet what I did and decide if they were logical.

Fryguy commented:
Ok thanks - the commit messages make complete sense.  I think we just need some specs here, and we are good to go.

kbrock commented:
update:
- added tests to validate `set_chargeback_report_options`

Fryguy commented on spec/models/chargeback_spec.rb:
```diff
@@ -0,0 +1,40 @@
+describe Chargeback do
+  # rpt.rpt_group     = @edit[:new][:rpt_group]
+  # rpt.rpt_type      = @edit[:new][:rpt_type]
+  # rpt.priority      = @edit[:new][:priority]
+  # rpt.categories    = @edit[:new][:categories]
+  # rpt.col_options   = @edit[:new][:col_options]
+
+  # rpt.order = @edit[:new][:sortby1].nil? ? nil : @edit[:new][:order]
```
@kbrock Should these comments stay?

kbrock replied:
no comment :(

Fryguy commented on spec/models/chargeback_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+  # called from ReportController::Reports::Editor
+  # must be called on a subclass of Chargeback
+  describe "#set_chargeback_report_options" do
```
Super-minor but since this is a class method:

```suggestion
  describe ".set_chargeback_report_options" do
```

Fryguy commented on spec/models/chargeback_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  # rpt.order = @edit[:new][:sortby1].nil? ? nil : @edit[:new][:order]
+
+  let(:rpt) { MiqReport.new(:db => chargeback_model.name, :headers => [], :col_formats => []) }
+  let(:chargeback_model) { ChargebackVm }
```
@kbrock Can you also add specs for ChargebackProject?

Fryguy replied:
Or is ChargebackVm also used for projects? (I see the project test below, so I was kinda surprised it worked?)

kbrock commented:
update:
- added `ChargebackContainerProject`
- other feedback

Fryguy commented on spec/models/chargeback_spec.rb:
```diff
@@ -0,0 +1,62 @@
+        expect(rpt.sortby).to eq(%w[start_date])
+      end
+
+      it "groups by project" do
+        chargeback_model.set_chargeback_report_options(rpt, "project", "Tag", "Group By", "UTF")
+
+        expect(rpt.cols).to eq(%w[start_date display_range project_name])
+        expect(rpt.col_order).to eq(%w[project_name display_range])
+        expect(rpt.sortby).to eq(%w[project_name start_date])
+      end
```
how can a ChargebackVm even have a project or be grouped by project? 

Since this one is unlikely to actually be able to occur since the user can't choose it I'm going to merge, but I feel like this particular case should be removed (or replaced with something VM specific)

Fryguy commented:
Backported to `petrosian` in commit 3f84b0b72323af05198f85ff473bbc00e04558ce.

```text
commit 3f84b0b72323af05198f85ff473bbc00e04558ce
Author: Jason Frey <fryguy9@gmail.com>
Date:   Fri Apr 7 17:49:57 2023 -0400

    Merge pull request #22445 from kbrock/chargeback_project_date_only
    
    Chargeback project date only
    
    (cherry picked from commit 6b8abb11c107ead2d3679b49672e6ec8eed5e695)
```
